### PR TITLE
Math related fixes

### PR DESCRIPTION
**Pull request by brunobar79**
Fixes APP-1570
Fixes RNBW-4788
Fixes APP-1587

## What changed (plus any additional context for devs)
- Alternative solution to the order of magnitude calculations that doesn't use log10 internally.
Added tests and confirmed an issue with the previous implementation where the order of magnitude was off by 1 for numbers < 1.
- Fix div by zero crash
- Capping input to 17 digits to prevent overflow & crahs

## Screen recordings / screenshots

https://github.com/rainbow-me/rainbow/assets/1247834/d9c45d7c-6151-450a-b418-e1e6fc575fb1

https://github.com/rainbow-me/rainbow/assets/1247834/3e52536a-6ca2-4940-8e5a-0503cd2eb6ef

https://github.com/rainbow-me/rainbow/assets/1247834/8e2fe0fa-6716-437f-9372-0d9416e29303

### CODE REVIEW

**Comment by walmat** [comment]
Rounding 50% of small fractions is back to being wonky on this branch.
## this branch
Screenshot 2024-06-12 at 11 58 13 AM
## develop
Screenshot 2024-06-12 at 12 00 16 PM

File: src/__swaps__/screens/Swap/components/SwapNumberPad.tsx [hunk 1]
```diff
@@ -48,6 +48,10 @@ export const SwapNumberPad = () => {
     const inputKey = focusedInput.value;
     const currentValue = SwapInputController.inputValues.value[inputKey].toString();
     const newValue = currentValue === '0' ? `${number}` : `${currentValue}${number}`;
+    // It will overflow with 18 digits
+    if (newValue.length > 17) {
+      return;
+    }
```
Comment (christianbaroni) [inline src/__swaps__/screens/Swap/components/SwapNumberPad.tsx:54]: Isn't this a string?

Also, aren't there valid input amounts that this will skip?

Comment (christianbaroni) [reply src/__swaps__/screens/Swap/components/SwapNumberPad.tsx:54]: (Meaning where does the overflow happen?)

**Comment by brunobar79** [comment]
> Rounding 50% of small fractions is back to being wonky on this branch.


Fixed. Pow: 

Screenshot 2024-06-12 at 5 31 14 PM
Screenshot 2024-06-12 at 5 31 14 PM

File: src/__swaps__/screens/Swap/components/SwapNumberPad.tsx (hunk 1 above)
Comment (brunobar79) [reply src/__swaps__/screens/Swap/components/SwapNumberPad.tsx]: Fixed and working up to 78 digits (max uint256 size)


https://github.com/rainbow-me/rainbow/assets/1247834/80ed05d1-bc5e-4b60-8c14-1300aeede46d

Comment (walmat) [reply src/__swaps__/screens/Swap/components/SwapNumberPad.tsx]: side note, but uhhhhhhh should we limit this to maybe a smaller amount than 78 😅

Comment (brunobar79) [reply src/__swaps__/screens/Swap/components/SwapNumberPad.tsx]: No

Comment (brunobar79) [reply src/__swaps__/screens/Swap/components/SwapNumberPad.tsx]: I'm pro letting anyone dump their entire token supply

**Review by walmat — APPROVED** [review]
code lgtm

**Review by jinchung — APPROVED** [review]
🌮

**Pull request merged**